nexus/src/app/background/tasks/saga_recovery.rs

```diff
@@ -0,0 +1,917 @@
+        )
+        .await;
+
+        // Process any newly-created sagas, adding them to our set of sagas
+        // to ignore during recovery.  We never want to try to recover a
+        // saga that was created within this Nexus's lifetime.
+        //
+        // We do this even if the previous step failed in order to avoid
+        // letting the channel queue build up.  In practice, it shouldn't
+        // really matter.
+        //
+        // But given that we're doing this, it's critical that we do it
+        // *after* having fetched the candidate sagas from the database.
+        // It's okay if one of these newly-created sagas doesn't show up in
+        // the candidate list (because it hadn't actually started at the
+        // point where we fetched the candidate list).  The reverse is not
+        // okay: if we did this step before fetching candidates, and a saga
+        // was immediately created and showed up in our candidate list, we'd
+        // erroneously conclude that it needed to be recovered when in fact
+        // it was already running.
+        let nstarted = self
+            .rest_state
+            .update_started_sagas(log, &mut self.sagas_started_rx);
+
+        match result {
+            Ok(db_sagas) => {
+                let plan = nexus_saga_recovery::Plan::new(
+                    log,
+                    &self.rest_state,
+                    db_sagas,
+                );
+                self.recovery_check_done(log, &plan).await;
+                let (execution, future) =
+                    self.recovery_execute(log, &plan).await;
+                self.rest_state.update_after_pass(&plan, &execution);
+                let last_pass_success =
+                    nexus_saga_recovery::LastPassSuccess::new(
+                        &plan, &execution,
+                    );
+                self.status.update_after_pass(&plan, execution, nstarted);
+                Some((future, last_pass_success))
+            }
+            Err(error) => {
+                self.status.update_after_failure(&error, nstarted);
+                None
+            }
+        }
+    }
+
+    /// Check that for each saga that we inferred was done, Steno agrees
+    ///
+    /// This is not strictly necessary because this should always be true.  But
+    /// if for some reason it's not, that would be a serious issue and we'd want
+    /// to know that.
+    async fn recovery_check_done(
+        &mut self,
+        log: &slog::Logger,
+        plan: &nexus_saga_recovery::Plan,
+    ) {
+        for saga_id in plan.sagas_inferred_done() {
+            match self.sec_client.saga_get(saga_id).await {
+                Err(_) => {
+                    self.status.ntotal_sec_errors_missing += 1;
+                    error!(
+                        log,
+                        "SEC does not know about saga that we thought \
+                        had finished";
+                        "saga_id" => %saga_id
```

Comment:
very tiny nit: should we maybe include the error in the log message?
```suggestion
                Err(error) => {
                    self.status.ntotal_sec_errors_missing += 1;
                    error!(
                        log,
                        "SEC does not know about saga that we thought \
                        had finished";
                        "saga_id" => %saga_id,
                        "error" => %error,
```
or, will these errors always be the same?

Reply:
The error type here is just `()`:
https://docs.rs/steno/latest/steno/struct.SecClient.html#method.saga_get

which is admittedly a little weird:
https://github.com/oxidecomputer/steno/blob/3f70f87aee5f7c6b5ff33eb70123f84b559cad61/src/sec.rs#L1258-L1261

but it can only mean that the saga was not found.

Reply:
Ah, gotcha. Carry on, then!